### PR TITLE
[torch] Enable GPU training

### PR DESCRIPTION
**Pull request by ShreyaR**

### CODE REVIEW

File: ludwig/models/ecd.py [hunk 1]
```diff
@@ -216,7 +217,20 @@ def train_loss(
             predictions,
             regularization_type: Optional[str] = None,
             regularization_lambda: Optional[float] = None
-    ):
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """
```
Comment (justinxzhao) [inline ludwig/models/ecd.py:221]: Nice!

File: ludwig/features/base_feature.py [hunk 2]
```diff
@@ -233,29 +217,27 @@ def update_metrics(self, targets: Tensor, predictions: Dict[str, Tensor]):
...
+            metric_fn = metric_fn.to(predictions[prediction_key].device)
             metric_fn.update(predictions[prediction_key].detach(), targets)
 
     def get_metrics(self):
         metric_vals = {}
         for metric_name, metric_onj in self.metric_functions.items():
             try:
-                metric_vals[metric_name] = metric_onj.compute(
-                ).detach().numpy().item()
+                metric_vals[metric_name] = metric_onj.compute().detach().cpu().numpy().item()
```
Comment (justinxzhao) [inline ludwig/features/base_feature.py:230]: This sequence `compute().detach().cpu().numpy().item()` appears both here and ecd.py. Does it make sense to put this in a shared function, i.e. `get_detached_cpu_metric(metric_obj)`?

File: ludwig/features/base_feature.py [hunk 3]
```diff
@@ -233,29 +217,27 @@ def update_metrics(self, targets: Tensor, predictions: Dict[str, Tensor]):
...
             prediction_key = metric_class.get_inputs()
+            # TODO(shreya): Metrics should ideally just move to the correct device
+            #  and not require the user to do this. This is a temporary fix. See
+            #  if this can be removed before merging the PR.
+            metric_fn = metric_fn.to(predictions[prediction_key].device)
             metric_fn.update(predictions[prediction_key].detach(), targets)
 
     def get_metrics(self):
         metric_vals = {}
         for metric_name, metric_onj in self.metric_functions.items():
```
Comment (justinxzhao) [inline ludwig/features/base_feature.py:228]: `metric_fn`?

File: ludwig/models/predictor.py [hunk 4]
```diff
@@ -136,7 +138,7 @@ def predict_single(self, model, batch):
 
     def _predict(self, model, batch):
         inputs = {
-            i_feat.feature_name: batch[i_feat.proc_column]
+            i_feat.feature_name: torch.from_numpy(batch[i_feat.proc_column]).to(self.device)
```
Comment (justinxzhao) [inline ludwig/models/predictor.py:141]: Could we document `_predict()` with the expected type of batch (a numpy array)?

File: ludwig/models/trainer.py [hunk 5]
```diff
@@ -569,10 +538,10 @@ def _is_valid_batch_size(batch_size):
                 low = batch_size
                 prev_batch_size = batch_size
                 try:
-                    # re-initialize model...
-                    model = LudwigModel.create_model(config, random_seed)
-                    self.train_for_tuning(
-                        model, training_set, batch_size, total_steps=3)
+                    # re-initalize model...
```
Comment (justinxzhao) [inline ludwig/models/trainer.py:541]: nit: initialized

File: ludwig/modules/metric_modules.py [hunk 6]
```diff
@@ -465,7 +465,7 @@ def __init__(self, **kwargs):
         super().__init__()
 
     def update(self, preds: Tensor, target: Tensor) -> None:
-        super().update(preds.detach(), target)
+        super().update(preds, target)
```
Comment (justinxzhao) [inline ludwig/modules/metric_modules.py:468]: What's the error if the `preds` tensor is detached?

File: ludwig/modules/tabnet_modules.py [hunk 7]
```diff
@@ -92,6 +90,11 @@ def __init__(
             output_size,
             output_size)
 
+        # Assign tensors to be used in forward pass
```
Comment (justinxzhao) [inline ludwig/modules/tabnet_modules.py:93]: Could we extend this comment with a higher-level statement about about why register_buffer is necessary for GPU training for these tensors specifically? Or perhaps a link to where this suggestion comes from?

**Review by justinxzhao — APPROVED** [review]
LGTM - just a few questions and small nits.

File: ludwig/modules/metric_modules.py (hunk 6 above)
Comment (ShreyaR) [reply ludwig/modules/metric_modules.py:468]: I tried rerunning the scripts that I was testing with and can't reproduce the error. I'll remove this change for now.

File: ludwig/modules/tabnet_modules.py (hunk 7 above)
Comment (ShreyaR) [reply ludwig/modules/tabnet_modules.py:93]: Done.

File: ludwig/features/base_feature.py (hunk 3 above)
Comment (ShreyaR) [reply ludwig/features/base_feature.py:228]: Done

File: ludwig/features/base_feature.py (hunk 2 above)
Comment (ShreyaR) [reply ludwig/features/base_feature.py:230]: Done.

File: ludwig/models/trainer.py (hunk 5 above)
Comment (ShreyaR) [reply ludwig/models/trainer.py:541]: I ended up removing this comment since we don't re-initialize the model here with the latest changes.

File: ludwig/models/ecd.py (hunk 1 above)
Comment (ShreyaR) [reply ludwig/models/ecd.py:221]: Ack!

**Comment by ShreyaR** [comment]
@justinxzhao uploaded a new patch responding to your comments!
